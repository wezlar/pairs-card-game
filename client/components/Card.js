import React, { Component } from 'react';
import styled from 'styled-components';

const CardWrapper = styled.div`
  position: relative;
  width: 150;
  height: 217px;
  margin-right: 10px;
  margin-bottom: 10px;
`;

const CardBody = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  box-shadow: 3px 3px 3px rgba(0,0,0,0.2);
  transform-style: preserve-3d;
  transition: transform .3s ease-in-out;
`;

const CardBack = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  z-index: 2;
  backface-visibility: hidden;
  background: #333;
`;

const CardFront = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  transition-delay: .15s;
  z-index: 1;
`;

class Card extends Component {
  render () {
    const { card: { image, name } } = this.props;

    const file = require(`../${image}`);

    return (
      <CardWrapper>
        <CardBody>
          <CardBack className="card__back">
            BACK { name }
          </CardBack>
          <CardFront className="card__face">
            <div dangerouslySetInnerHTML={ { __html: file } } />
          </CardFront>
        </CardBody>
      </CardWrapper>
    );
  }
};

export default Card;