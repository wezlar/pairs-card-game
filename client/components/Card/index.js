import React, { memo } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const cardWidth = 138;
const cardHeight = 200;

const CardWrapper = styled.div`
  position: relative;
  width: ${cardWidth + 10}px;
  height: ${cardHeight + 10}px;
  padding-top: 5px;
  padding-right: 5px;
  padding-bottom: 5px;
  padding-left: 5px;
  perspective: 800px;
  overflow: hidden;
`;

const CardBody = styled.div`
  position: absolute;
  width: ${cardWidth}px;
  height: ${cardHeight}px;
  box-shadow: 3px 3px 3px rgba(0,0,0,0.2);
  transform-style: preserve-3d;
  transition: transform .3s ease-in-out;
  border-radius: 5px;
  transform: rotateY(0);
`;

const CardBack = styled.div`
  position: absolute;
  width: ${cardWidth}px;
  height: ${cardHeight}px;
  z-index: 2;
  backface-visibility: hidden;
  background-color:#fff;
  background-image: radial-gradient(closest-side, transparent 98%, red 99%),
    radial-gradient(closest-side, transparent 98%, black 99%);
  background-size: 30px 30px;
  background-position: 0 0, 15px 15px;
  border:solid 10px #fff;
  border-radius: 5px;
`;

const CardFront = styled.div`
  position: absolute;
  width: ${cardWidth}px;
  height: ${cardHeight}px;
  transition-delay: .15s;
  z-index: 0;
  border-radius: 5px;
`;

const SVGWrapper = styled.div`
  svg {
    width: ${cardWidth}px;
    height: ${cardHeight}px;
  }
`;

export function Card (props = {}) {
  const { 
    card: {
      image, 
      isFlipped,
      hasMatched,
    },
    onClick,
  } = props;

  const file = image ? require(`!!raw-loader!../../${image}`) : '';
  const flippedStyle = isFlipped ? { transform: 'rotateY(-180deg)' } : {};
  const flippedFrontStyle = isFlipped ? { transform: 'rotateY(180deg)' } : {};
      
  const hasMatchedStyle = hasMatched 
    ? { boxShadow: '0 0 1px 2px #099f09' } 
    : { boxShadow: '0 0 1px 2px #00000000' };

  const cardFrontStyle = Object.assign({}, flippedFrontStyle, hasMatchedStyle);

  const cardFront = () =>
    <CardFront className="card__front" style={cardFrontStyle}>
      <SVGWrapper dangerouslySetInnerHTML={ { __html: file } } />
    </CardFront>;
  
  const cardBack = () => 
    <CardBack className="card__back" />;

  const cardBody = () => 
    <CardBody className="card__body" style={flippedStyle}>
      {cardBack()}
      {cardFront()}
    </CardBody>;

  // TODO: add hover state that slightly moves the card in opposite direction
  return <CardWrapper className="card__wrapper" onClick={onClick}>
      {cardBody()}
    </CardWrapper>;
};

Card.propTypes = {
  card: PropTypes.object,
  onClick: PropTypes.func,
};

Card.defaultProps = {
  card: {},
};

export default memo(Card);