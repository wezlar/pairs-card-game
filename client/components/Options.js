import React, { memo } from 'react'
import PropTypes from 'prop-types'

export function Score (props = {}) {
  const { score } = props

  return (
    <div>
      <div className="score">
        <span className="label">Score: </span>
        <span className="score">{score}</span>
      </div>
      <div className="start-new-game">
        <button>New Game</button>
      </div>
    </div>
  )
}

Score.propTypes = {
  score: PropTypes.number,
}

export default memo(Score)
