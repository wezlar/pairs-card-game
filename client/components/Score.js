import React, { memo } from 'react'
import PropTypes from 'prop-types'

export function Score (props = {}) {
  const { score } = props

  return (
    <div>
      <div className="score">
        <span class="label">Score: </span>
        <span class="score">{score}</span>
      </div>
      <div className="start-new-game">
        <button>New Game</button>
      </div>
    </div>
  )
}

Score.propTypes = {
  score: PropTypes.number,
}

export default memo(Score)
