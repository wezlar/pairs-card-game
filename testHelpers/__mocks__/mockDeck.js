const mockDeck = [
  {
    cardIndex: 0,
    hasMatched: false,
    image: 'images/svg-cards/2_of_hearts.svg',
    isFlipped: false,
    name: '2 of hearts',
  },
  {
    cardIndex: 1,
    hasMatched: false,
    image: 'images/svg-cards/3_of_hearts.svg',
    isFlipped: false,
    name: '3 of hearts',
  },
  {
    cardIndex: 2,
    hasMatched: false,
    image: 'images/svg-cards/4_of_hearts.svg',
    isFlipped: false,
    name: '4 of hearts',
  },
  {
    cardIndex: 0,
    hasMatched: false,
    image: 'images/svg-cards/2_of_hearts.svg',
    isFlipped: false,
    name: '2 of hearts',
  },
  {
    cardIndex: 1,
    hasMatched: false,
    image: 'images/svg-cards/3_of_hearts.svg',
    isFlipped: false,
    name: '3 of hearts',
  },
  {
    cardIndex: 2,
    hasMatched: false,
    image: 'images/svg-cards/4_of_hearts.svg',
    isFlipped: false,
    name: '4 of hearts',
  },
];

export default mockDeck;