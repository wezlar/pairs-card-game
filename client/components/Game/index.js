import React, { Component } from 'react';
import styled from 'styled-components';

import Card from '../Card';

const DeckWrapper = styled.div`
  position: relative;
  max-width: 1000px;
  margin-left: auto;
  margin-right: auto;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  padding: 20px;
`;

class Game extends Component {
  constructor (props) {
    super(props);
    this.returnState = null;
  }
  
  cardFlip (index) {
    const { deck, lastCardSelected } = this.props.game;
    const { isFlipped } = deck[index];

    if (isFlipped) {
      // I decided that I don't want people unflipping cards
      return;
    } 

    this.props.updateScore(1);
    
    // flip card
    this.props.cardFlip(index);
    
    // if first card selected
    if (lastCardSelected.name === '') {
      this.props.updateLastCard(index, deck[index].name);
      return;
    }
    
    // second card selected.
    if (lastCardSelected.name === deck[index].name) {
      this.props.cardMatched(index);
      this.props.cardMatched(lastCardSelected.index);
      this.props.updateCardsMatched(2);
    } else {
      this.returnState = setInterval(() =>{
        clearInterval(this.returnState);
        this.props.cardFlip(lastCardSelected.index);
        this.props.cardFlip(index);
      }, 1000);
    }

    // clear previous selection
    this.props.updateLastCard(null, '');
  }
  
  render () {
    const { deck } = this.props.game;

    return (
      <div className='game-wrapper'>
        <DeckWrapper>
          {deck.map((card, i) => <Card key={i} card={card} onClick={() => this.cardFlip(i)} />) }
        </DeckWrapper>
      </div>
    );
  }
};

export default Game;