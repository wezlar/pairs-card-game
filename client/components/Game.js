import React, { Component } from 'react';
import styled from 'styled-components';

import Card from './Card';
import Options from './Options';

import newGame from '../utils/newGame';

const DeckWrapper = styled.div`
  position: relative;
  max-width: 1000px;
  margin-left: auto;
  margin-right: auto;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
`;

class Game extends Component {
  constructor (props) {
    super(props);
    this.returnState = null;
    this.startNewGame = this.startNewGame.bind(this);
  }

  componentDidMount () {
    this.startNewGame();
  }

  startNewGame () {
    const { game: { numberOfCards }, cards } = this.props;

    console.log('Start new game function called');
    const newDeck = newGame(cards, numberOfCards);
    this.props.addNewDeck(newDeck);

  }
  
  cardFlip (index) {
    const { deck, lastCardSelected } = this.props.game;
    const { isFlipped } = deck[index];

    this.props.updateScore(1);

    if (isFlipped) {
      // I decided that I don't want people unflipping cards
      return;
    } 

    // flip card
    this.props.cardFlip(index);
    
    // if first card selected
    if (lastCardSelected.name === '') {
      this.props.updateLastCard(index, deck[index].name);
      return;
    }
    
    // second card selected.
    if (lastCardSelected.name === deck[index].name) {
      this.props.cardMatched(index);
      this.props.cardMatched(lastCardSelected.index);
      this.props.updateCardsMatched(2);
    } else {
      this.returnState = setInterval(() =>{
        clearInterval(this.returnState);
        this.props.cardFlip(lastCardSelected.index);
        this.props.cardFlip(index);
      }, 1000);
    }

    // clear previous selection
    this.props.updateLastCard(null, '');
  }
  
  render () {
    const { deck, score, numberOfCards, cardsMatched } = this.props.game;

    return (
      <div>
        <Options 
          isGameComplete={numberOfCards === cardsMatched}
          score={score} 
          startNewGame={this.startNewGame} />
        <DeckWrapper>
          {deck.map((card, i) => <Card key={ i } card={ card } onClick={() => this.cardFlip(i)} />) }
        </DeckWrapper>
      </div>
    );
  }
};

export default Game;