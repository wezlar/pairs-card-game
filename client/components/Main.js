import React, { Component } from 'react';
import { Link } from 'react-router';

import Options from './Options';

import newGame from '../utils/newGame';

class Main extends Component {
  constructor (props) {
    super(props);
    this.startNewGame = this.startNewGame.bind(this);
  }

  componentDidMount () {
    this.startNewGame();
  }

  startNewGame () {
    const { game: { numberOfCards, deck }, cards } = this.props;

    // loop through current deck and set flip to false so
    // new deck does not show before render
    deck.forEach((element, i) => {
      if (element.isFlipped) {
        console.log('Flipping ' + i);
        this.props.cardFlip(i);
      }
    });

    const newDeck = newGame(cards, numberOfCards);
    this.props.addNewDeck(newDeck);
  }
  
  render () {
    const { deck, score, numberOfCards, cardsMatched } = this.props.game;

    return (
      <div>
        <header>
          <h1>
            <Link to="/">Pairs</Link>
          </h1>
          <Options 
            isGameComplete={numberOfCards === cardsMatched}
            score={score} 
            startNewGame={this.startNewGame} />
        </header>
        { React.cloneElement(this.props.children, this.props) }
      </div>
    );
  }
};

export default Main;