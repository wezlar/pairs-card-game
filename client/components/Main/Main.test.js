import React from 'react';
import { mount } from 'enzyme';
import { MockSimpleChild } from '@jest-helpers/components';
import mockDeck from '../../../testHelpers/__mocks__/mockDeck';
import Component from './Main';

const mockProps = {
  game: {
    numberOfCards: 10,
    score: 0,
    cardsMatched: 0,
    deck: mockDeck,
  },
};

describe(`<Main />`, () => {
  test(`It doesn't blow up` , () => {
    const tree = mount(<Component {...mockProps}><MockSimpleChild /></Component>);
    expect(tree.length).toEqual(1);
  });

  test(`Renders correctly`, () => {
    const main = mount(<Component {...mockProps}><MockSimpleChild /></Component>);
    expect(main.debug()).toMatchSnapshot();
  });
});