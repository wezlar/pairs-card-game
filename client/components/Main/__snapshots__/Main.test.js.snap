// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main /> Renders correctly 1`] = `
"<Main game={{...}}>
  <div>
    <header>
      <h1>
        <Link to=\\"/\\" onlyActiveOnIndex={false} style={{...}}>
          <a style={{...}} onClick={[Function: bound handleClick]}>
            Pairs
          </a>
        </Link>
      </h1>
      <Options isGameComplete={false} score={0} startNewGame={[Function: bound startNewGame]}>
        <div className=\\"options\\">
          <div className=\\"score\\">
            <span className=\\"label\\">
              Score: 
            </span>
            <span className=\\"score-value\\">
              0
            </span>
          </div>
          <div className=\\"start-new-game\\">
            <button className=\\"start-new-game__button\\" onClick={[Function: bound startNewGame]}>
              New Game
            </button>
          </div>
        </div>
      </Options>
    </header>
    <MockSimpleChild game={{...}}>
      <div className=\\"\\" />
    </MockSimpleChild>
  </div>
</Main>"
`;
